gridcomps/History/Sampler/MAPL_StationSamplerMod.F90

```diff
@@ -73,155 +101,226 @@ function new_StationSampler_readfile (filename,nskip_line, rc) result(sampler)
     !      ["name_short lat lon elev name_full"]
     !
 
-    open(newunit=unit, file=trim(filename), form='formatted', &
-         access='sequential', status='old', _IOSTAT)
-    ios=0
-    nstation=0
-    nskip=0
-    if (present(nskip_line)) then
-       nskip=nskip_line
-    end if
-    if (nskip>0) then
-       do i=1, nskip
-          read(unit, *)
-       end do
-    end if
-    read(unit, '(a100)', IOSTAT=ios) line
-    call count_substring(line, ',', ncount, _RC)
-    con1= (ncount>=2 .AND. ncount<=4).OR.(ncount==0)
-    _ASSERT(con1, 'string sequence in Aeronet file not supported')
-    if (ncount==0) then
-       seq='AFFFA'
-    elseif (ncount==2) then
-       seq='AFF'
-    elseif (ncount==3) then
-       seq='AFFF'
-    elseif (ncount==4) then
-       CH1=line(1:1)
-       con1= (CH1>='a'.AND.CH1<='z').OR.(CH1>='A'.AND.CH1<='Z')
-       con2= CH1>='0'.AND.CH1<='9'
-       if (con1) then
-          seq='AIFFF'
-       else
-          if (con2) then
-             seq='IAFFF'
+    lgr => logging%get_logger('HISTORY.sampler')
+    if ( MAPL_AM_I_ROOT() ) then
+       open(newunit=unit, file=trim(filename), form='formatted', &
+            access='sequential', status='old', _IOSTAT)
+       ios=0
+       nstation=0
+       nskip=0
+       if (present(nskip_line)) then
+          nskip=nskip_line
+       end if
+       if (nskip>0) then
+          do i=1, nskip
+             read(unit, *)
+          end do
+       end if
+       read(unit, '(a100)', IOSTAT=ios) line
+       call count_substring(line, ',', ncount, _RC)
+       con1= (ncount>=2 .AND. ncount<=4).OR.(ncount==0)
+       _ASSERT(con1, 'string sequence in Aeronet file not supported')
+       if (ncount==0) then
+          seq='AFFFA'
+       elseif (ncount==2) then
+          seq='AFF'
+       elseif (ncount==3) then
+          seq='AFFF'
+       elseif (ncount==4) then
+          CH1=line(1:1)
+          con1= (CH1>='a'.AND.CH1<='z').OR.(CH1>='A'.AND.CH1<='Z')
+          con2= CH1>='0'.AND.CH1<='9'
+          if (con1) then
+             seq='AIFFF'
           else
-             _ASSERT(.false., 'string sequence in Aeronet file not supported')
+             if (con2) then
+                seq='IAFFF'
+             else
+                _ASSERT(.false., 'string sequence in Aeronet file not supported')
+             end if
           end if
        end if
-    end if
 
-    rewind(unit)
-    if (nskip>0) then
-       do i=1, nskip
-          read(unit, *)
+       rewind(unit)
```

Comment:
At first glance this looks like it is outside the AM_I_ROOT conditional.  Are all of the processes doing the reading?

More generally, this is an awfully long (and nested conditional).   I suggest a follow up PR to keep the block size smaller.